downstairs/src/extent_inner_raw.rs

```diff
@@ -0,0 +1,2375 @@
+                * BLOCK_CONTEXT_SLOT_SIZE_BYTES
+    }
+
+    /// Number of blocks in the extent file
+    fn block_count(&self) -> u64 {
+        self.extent_size.value
+    }
+
+    /// Returns the byte offset of the `active_context` bitpacked array
+    fn active_context_offset(&self) -> u64 {
+        self.supplementary_data_offset()
+            + self.block_count() * 2 * BLOCK_CONTEXT_SLOT_SIZE_BYTES
+    }
+
+    fn active_context_size(&self) -> u64 {
+        (self.block_count() + 7) / 8
+    }
+
+    fn metadata_offset(&self) -> u64 {
+        self.active_context_offset() + self.active_context_size()
+    }
+
+    /// Number of bytes in each block
+    fn block_size(&self) -> u64 {
+        self.extent_size.block_size_in_bytes() as u64
+    }
+
+    fn get_metadata(&self, file: &File) -> Result<OnDiskMeta, CrucibleError> {
+        let mut buf = [0u8; BLOCK_META_SIZE_BYTES as usize];
+        let offset = self.metadata_offset();
+        nix::sys::uio::pread(file.as_raw_fd(), &mut buf, offset as i64)
+            .map_err(|e| {
+                CrucibleError::IoError(format!("reading metadata failed: {e}"))
+            })?;
+        let out: OnDiskMeta = bincode::deserialize(&buf)
+            .map_err(|e| CrucibleError::BadMetadata(e.to_string()))?;
+        Ok(out)
+    }
+
+    fn write_context_slots_contiguous<'a, I>(
+        &self,
+        file: &File,
+        block_start: u64,
+        iter: I,
+        slot: ContextSlot,
+    ) -> Result<(), CrucibleError>
+    where
+        I: Iterator<Item = Option<&'a DownstairsBlockContext>>,
+    {
+        let mut buf = self.buf.take();
+        buf.clear();
+
+        for block_context in iter {
+            let n = buf.len();
+            buf.extend([0u8; BLOCK_CONTEXT_SLOT_SIZE_BYTES as usize]);
+            let d = block_context.map(|b| OnDiskDownstairsBlockContext {
+                block_context: b.block_context,
+                on_disk_hash: b.on_disk_hash,
+            });
+            bincode::serialize_into(&mut buf[n..], &d).unwrap();
+        }
+        let offset = self.context_slot_offset(block_start, slot);
+        nix::sys::uio::pwrite(file.as_raw_fd(), &buf, offset as i64).map_err(
+            |e| {
+                CrucibleError::IoError(format!(
+                    "writing context slots failed: {e}"
+                ))
+            },
+        )?;
+        self.buf.set(buf);
+        Ok(())
+    }
+
+    fn read_context_slots_contiguous(
+        &self,
+        file: &File,
+        block_start: u64,
+        block_count: u64,
+        slot: ContextSlot,
+    ) -> Result<Vec<Option<DownstairsBlockContext>>, CrucibleError> {
+        let mut buf = self.buf.take();
+        buf.resize((BLOCK_CONTEXT_SLOT_SIZE_BYTES * block_count) as usize, 0u8);
+
+        let offset = self.context_slot_offset(block_start, slot);
+        nix::sys::uio::pread(file.as_raw_fd(), &mut buf, offset as i64)
+            .map_err(|e| {
```

Comment:
We don't call `buf.clear()` here, 
Is their any chance the `pread` will return with a partially filled buffer?
It's only us who use the buffer in the downstairs, so there should not be
a chance for someone to get old data from a previous read, but just
wondering here if that is a possibility.

Reply:
Good question!  Quote [the standard](https://pubs.opengroup.org/onlinepubs/9699919799/functions/pread.html),

> The value returned may be less than nbyte if the number of bytes left in the file is less than nbyte, if the read() request was interrupted by a signal, or if the file is a pipe or FIFO or special file and has fewer than nbyte bytes immediately available for reading. For example, a read() from a file associated with a terminal may return one typed line of data.

The only concerning part here is "interrupted by a signal"; I think everything else isn't relevant.  Do you think we need to wrap every `pread(v)/pwrite(v)` in a little loop?  Right now, the SQLite backend is using `pread` without any special handling.

Reply:
I've added some handling of this in 2a5ed27c32caa2922e620a5e5da7cb467ceeff4f

Specifically, `pread` and `pwrite` will now loop until the full buffer is read / written, and `preadv` / `pwritev` will return an `IoError` (because I didn't want to write that handling logic yet).

Reply:
Looks good, and I think not having the error retry loop is fine as I don't believe we have ever seen this error, and if we do, it's probably something catastrophic going on anyway.